test/filters/test_filters.cpp

```diff
@@ -1399,16 +1399,16 @@ TEST (VoxelGridMinPoints, Filters)
   // Verify 2 clusters (0.11 and 0.31) passed threshold and verify their location and color
   EXPECT_EQ (out_pc->points.size (), 2);
 
-  EXPECT_NEAR (out_pc->at(0).x, 0.11, 1e-2);
-  EXPECT_NEAR (out_pc->at(0).y, 0.11, 1e-2);
-  EXPECT_NEAR (out_pc->at(0).z, 0.11, 1e-2);
-  EXPECT_NEAR (out_pc->at(0).r, 200, 1);
+  EXPECT_NEAR (out_pc->at(0).x, input->at(1).x, 1e-2);
```

Comment:
Not input, but output and not expect_near, but expect_eq

Reply:
I think there is still a risk of floating point rounding error differences from the 2 paths, I compared to output and lowered the comparison epsilon below the magnitude of the applied offsets.

Reply:
Exact same operations should happen, right? If so, then the output would be the same on all standard floats (same ops in same order give same result)

Reply:
The code is slightly different, for example in the downsample all data case for PointCloudXYZRGB:
```
      CentroidPoint<PointT> centroid;

      // fill in the accumulator with leaf points
      for (unsigned int li = first_index; li < last_index; ++li)
        centroid.add ((*input_)[index_vector[li].cloud_point_index]);  

      centroid.get (output[index]);
```

vs. PCLPointCloud2

```
      Eigen::VectorXf centroid = Eigen::VectorXf::Zero (centroid_size);
      // fill in the accumulator with leaf points
      for (index_t li = first_index; li < last_index; ++li) {
        std::size_t point_offset = index_vector[li].cloud_point_index * input_->point_step;
        // ---[ RGB special case
        // fill extra r/g/b centroid field
        if (rgba_index >= 0)
        {
            pcl::RGB rgb;
            memcpy (&rgb, &input_->data[point_offset + input_->fields[rgba_index].offset], sizeof (RGB));
            temporary[centroid_size-4] = rgb.r;
            temporary[centroid_size-3] = rgb.g;
            temporary[centroid_size-2] = rgb.b;
            temporary[centroid_size-1] = rgb.a;
        }
        // Copy all the fields
        for (std::size_t d = 0; d < input_->fields.size (); ++d)
          memcpy (&temporary[d], &input_->data[point_offset + input_->fields[d].offset], field_sizes_[d]);
        centroid += temporary;
      }
      centroid /= static_cast<float> (last_index - first_index);
      
      std::size_t point_offset = index * output.point_step;
      // Copy all the fields
      for (std::size_t d = 0; d < output.fields.size (); ++d)
        memcpy (&output.data[point_offset + output.fields[d].offset], &centroid[d], field_sizes_[d]);

      // ---[ RGB special case
      // full extra r/g/b centroid field
      if (rgba_index >= 0) 
      {
        float r = centroid[centroid_size-4], g = centroid[centroid_size-3], b = centroid[centroid_size-2], a = centroid[centroid_size-1];
        int rgb = (static_cast<int> (a) << 24) | (static_cast<int> (r) << 16) | (static_cast<int> (g) << 8) | static_cast<int> (b);
        memcpy (&output.data[point_offset + output.fields[rgba_index].offset], &rgb, sizeof (float));
      }
```

Its possible those two blocks of code end up being the same opertations in the same order, but I don't have time to verify.  I could change the unit test to check strict equals, but it runs that risk of failing on an untested setup if it turns out the operations are not quite the same.  The PCLPointCloud2 code is re-used from the code block that existed before for the downsample all data case (just now factoring in the limited voxels that passed the threshold test)

Reply:
You're right.